### PR TITLE
Start of PyKEEN integration

### PR DESCRIPTION
**Pull request by Mackofi**
I have implemented the basic features PyKEEN provides (according to the [design document](https://docs.google.com/document/d/1LQFC0kzm5ac508XQJcTki3O9tH3X4YunWlb-uFf3Ds4)), but I am sure there are things that can and will be improved down the line.

Sadly we discovered that loading embeddings into PyKEEN models from other sources (as described in #29) is not possible the way we wanted it, as PyKEEN only supports edge types. So that feature will have to be skipped, unless we find a reasonable replacement.

I have also provided a workspace demonstrating the usage of PyKEEN, and I am currently working on an example that involves another machine learning task using the embeddings extracted from the trained PyKEEN model. But that is very much not ready yet.

The workspace is still work in progress and I will continue to iterate on it, but right now everything is in a good enough state for a draft.

### CODE REVIEW

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    if "model" in bundle.other:
+        model = bundle.other["model"]
+    else:
+        sampler = None
+        if model is PyKEENModel.RGCN and training_approach == TrainingType.sLCWA:
+            # Currently RGCN is the only model that requires a sampler and only when using sLCWA
+            sampler = "schlichtkrull"
+        model = model.to_class()
+    training_set = TriplesFactory.from_labeled_triples(
+        bundle.dfs["triples_train"][["head", "relation", "tail"]].values,
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:184]: Take the train/test table names as parameters.

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        entity_embeddings = actual_model.entity_representations[0]().detach().cpu()
+
+    if entity_embeddings is None:
+        raise AttributeError(
+            f"Cannot extract entity embeddings from model type: {type(actual_model)}. "
+            f"Available attributes: {[attr for attr in dir(actual_model) if not attr.startswith('_')]}"
+        )
+
+    bundle.other[node_embedding_name] = entity_embeddings
+    bundle.other["entity_to_index"] = {entity: idx for idx, entity in enumerate(entity_labels)}
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:313]: Instead of dicts, put these in a DataFrame. If we already have (had) a `nodes` table, these could be inserted into that as new columns.

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+    return bundle
+
+
+@op("Extract embeddings from PyKEEN model")
+def extract_from_pykeen(bundle: core.Bundle, *, node_embedding_name: str, edge_embedding_name: str):
+    bundle = bundle.copy()
+    model = bundle.other["model"]
+    triples = TriplesFactory.from_labeled_triples(
+        bundle.dfs["triples_train"][["head", "relation", "tail"]].values,
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:290]: Take table name as a parameter.

Must this be the same table as the one used for the training? If it can only be used with that, then perhaps it's best to extract these at the end of training, as part of the training box.

File: examples/pykeen-testing.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+
+@op("Load Embedding into PyKEEN model")
+def load_pykeen_embeddings(bundle: core.Bundle, *, model: str):
+    """
+    Assuming embeddings are torch tensors
+    Entity names has to be a file with comma seperated entity names
+    """
+    node_embeddings = bundle.other["node_embedding"]
+    edge_embeddings = bundle.other["edge_embedding"]
```
Comment (darabos) [inline examples/pykeen-testing.py:80]: Stuff that is per-node or per-edge should be stored in the nodes/edges DataFrames.

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    bundle.dfs["training"] = pd.DataFrame({"training_loss": result.losses})
+    bundle.other["model"] = result.model
+
+    return bundle
+
+
+@op("Triples prediction")
+def triple_predict(bundle: core.Bundle, *, table_name: str = "triples_val"):
+    bundle = bundle.copy()
+    model = bundle.other.get("model")
```
[the remark this answers is not in the document]
Comment (Mackofi) [reply lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py]: I think I have done it correctly, renamed the `ModelName` into `PyTorchModelName` and the type into `pytorch-model`, renamed the symbol, so it should've replaced everything using the old naming system. 
Also created a `PyKEENModelName` and `PyKEENModelWrapper` that implements this functionality, and it works flawlessly.

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+            create_inverse_triples=req_inverse_triples(model),
+        ),
+    )
+    bundle.dfs["pred"] = pred_annotated.df
+
+    return bundle
+
+
+@op("Full prediction", slow=True)
+def full_predict(bundle: core.Bundle, *, k: int = None):
```
[the remark this answers is not in the document]
Comment (Mackofi) [reply lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py]: This has been fixed in #44.

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py (hunk 2 above)
Comment (darabos) [reply lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:373]: > Should we convert node/edge embeddings into dataframes from tensors, just so we can then later convert them back into tensors for machine learning use?

Yes.

- Often we can't keep data in tensors because it's too much. Usually neural networks process data in batches sized so that each batch fills up the GPU. The rest of the data is kept on disk or in system memory.
- Tensors often have gradients attached to them which we probably don't want to store.
- Putting the tensors into the DataFrame shouldn't be inefficient. LynxKite uses `cudf.pandas` if `NX_CUGRAPH_AUTOCONFIG` is `true`. In that case I would hope that somehow the various libraries involved in this do a good job. If they don't, that will have to be fixed, but I don't think it should be a design constraint on us.
- This is needed to be compatible with the rest of LynxKite. Per-node/per-edge data is stored in the DataFrames.

> Also I have done the following with `entity_to_index`: I created a `nodes` table during the importing of datasets and then saved the `label`s and `id`s into two columns. Same with relations/edges. But there I saved `source, relation, target, relation_id`.

Awesome!

Comment (Mackofi) [reply lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py]: Alright, I understand, it was just a bit weird to me, but with this I fully understand the reason. Implemented it!

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    # Validation -----------
+    triples = ds.validation.mapped_triples
+    df_val = mapped_triples_to_df(
+        triples=triples,
+        entity_to_id=ds.entity_to_id,
+        relation_to_id=ds.relation_to_id,
+    )
+
+    bundle = core.Bundle()
+    bundle.dfs["triples_train"] = df_train
```
[the remark this answers is not in the document]
Comment (Mackofi) [reply lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py]: I've looked into this further, and found the following:
- Datasets do not have entity properties, so we don't need to worry about that.
- The "Visualize graph" box works quite well with the current `nodes` and `edges` tables, but the edges have to be transformed from their labels into their ids first.
- Aren't relations in a bundle in the context of databases? How would that work with edge types?
- I have not looked into relation selections but filtering is more of a general task in my opinion and should be done in a separate box/SQL query, but if we need to do it in the "Visualize graph" box then I would be able to do that.
- And the last point should already work (with the `edges_train`, `edges_test`, `edges_val` tables), if transformed from their labels to ids, and if we could select which tables are the nodes and edges located at before graph visualization.

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 8]
```diff
@@ -0,0 +1,633 @@
...
+        optimizer=optimizer_type,
+        evaluator=None,
+        random_seed=42,
+        stopper="early",
+        stopper_kwargs=dict(
+            frequency=2,
+            patience=6,
+            relative_delta=0.001,
+            metric="hits@k",
+        ),
```
Comment (Mackofi) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:380]: Right now there are hard coded, but it would be cool if I (or someone else) came up with a solution to this that does not include just adding 4-5 parameters to the function call...

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 9]
```diff
@@ -163,18 +164,34 @@ class PyKEENModel(str, enum.Enum):
     TransR = "TransR"
     TuckER = "TuckER"
 
-    def to_class(self, triples_factory=None, embedding_dim=None):
+    def to_class(self, triples_factory: TriplesFactory, embedding_dim: int) -> models.Model:
         return getattr(models, self.value)(
             triples_factory=triples_factory,
             embedding_dim=embedding_dim,
+            random_seed=42,
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:171]: Let's make the seed controllable by the user! Just make it a parameter of the box. I love having seed parameters — this makes it easy to re-run or not re-run the box.

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 10]
```diff
@@ -192,8 +209,44 @@ def __getattr__(self, name):
     def __str__(self):
         return str(self.model)
 
+    def __getstate__(self):
+        state = {
+            "trained": getattr(self, "trained", False),
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:216]: Why not just

```suggestion
            "trained": self.trained,
```

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 11]
```diff
@@ -192,8 +209,44 @@ def __getattr__(self, name):
     def __str__(self):
         return str(self.model)
 
+    def __getstate__(self):
+        state = {
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:215]: Instead of listing everything, maybe consider:

```
state = dict(self.__dict__)
del state["model"]
```

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/core.py [hunk 12]
```diff
@@ -42,17 +42,28 @@
...
         "format": "dropdown",
-        "metadata_query": "[].other.*[] | [?type == 'model'].key",
+        "metadata_query": "[].other.*[] | [?type == 'pytorch-model'].key",
     },
 ]
 """A type annotation to be used for parameters of an operation. ModelName is
 rendered as a dropdown in the frontend, listing the models in the Bundle.
 The model name is passed to the operation as a string."""
 
+PyKEENModelName = typing.Annotated[
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/core.py:56]: Please move this to the PyKEEN code file. (You could also move PyTorchModelName to the PyTorch code.)

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/lynxkite_ops.py [hunk 13]
```diff
@@ -135,9 +135,9 @@ def import_graphml(*, filename: str):
 
 @op("Graph from OSM", slow=True)
 def import_osm(*, location: str):
-    import osmnx as ox
+    from osmnx import graph
 
-    return ox.graph.graph_from_place(location, network_type="drive")
+    return graph.graph_from_place(location, network_type="drive")
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/lynxkite_ops.py:140]: Revert this?

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py [hunk 14]
```diff
@@ -0,0 +1,633 @@
...
+
+    metrics = metrics_str.split(",")
+    metrics_df = pd.DataFrame(columns=["metric", "score"])
+
+    for metric in metrics:
+        metric = metric.strip()
+        try:
+            score = evaluated.get_metric(metric)
+        except Exception as e:
+            raise Exception(f"Possibly unknown metric: {metric}\nError: {e}")
```
Comment (darabos) [inline lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:626]: Use [exception chaining](https://docs.python.org/3/tutorial/errors.html#exception-chaining):

```suggestion
            raise Exception(f"Possibly unknown metric: {metric}") from e
```

**Review by darabos — APPROVED** [review]
Awesome, thanks!

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py (hunk 10 above)
Comment (Mackofi) [reply lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:216]: I think that was co-pilot tab completion just seeing getattr in a lot of places and deciding to use that here too. I have replaced this whole dict with your suggestion of
```python
state = dict(self.__dict__)  
del state["model"] 
```

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py (hunk 9 above)
Comment (Mackofi) [reply lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py:171]: Done, I did it so that the training run uses the same seed as the model.
